File: MATLAB/CircularCylinderScatteringFEM.m
clear
clc

lambda = 1;
f = 300*10^6;
k = (2*pi)/lambda;
a = lambda/8;
R = a + lambda/2;
E0 = 1;
m0 = 1.256637061*10^(-6);
e0 = 8.854187817* 10^(-12);  

alpha = -(1/(2*R) + 1i*k);
omega = ((2*pi*f)^2);

% Geometry Description
gd = [1 1; 0 0; 0 0; R a];
ns = [82 82; 49 50];
dl = decsg(gd, 'R1-R2', ns);

% Create Mesh
[p, e, t] = initmesh(dl);
[p, e, t] = refinemesh(dl, p, e, t);
[p, e, t] = refinemesh(dl, p, e, t);
[p, e, t] = refinemesh(dl, p, e, t);
Num_nodes = size(p, 2);
Num_elements = size(t, 2);

pdeplot(p, e, t);

% Define known and unknown fields
node_id = ones(Num_nodes, 1);
Ez = zeros(Num_nodes, 1);
Num_edges = size(e, 2);
for ie = 1:Num_edges
    region1 = e(6, ie);
    region2 = e(7, ie);
    nodes(1:2) = e(1:2, ie);
    x(1:2) = p(1, nodes(1:2));
    y(1:2) = p(2, nodes(1:2));
    radius(1:2) = sqrt(x(1:2).^2 + y(1:2).^2);

    if (region1 == 0 || region2 == 0)
        
        if (radius(1) < R/2) 
            Ez(nodes(1)) = -E0*exp(-1i*2*pi*p(1,nodes(1)));
            node_id(nodes(1)) = 0;
        else
            node_id(nodes(1)) = 2;
        end
        if (radius(2) < R/2) 
            Ez(nodes(2)) = -E0*exp(-1i*2*pi*p(1,nodes(2)));
            node_id(nodes(2)) = 0;
        else
            node_id(nodes(2)) = 2;
        end
    end
end

% Indexes of unknowns
index = zeros(Num_nodes, 1);
counter = 0;
for in = 1:Num_nodes
    if (node_id(in) ~= 0)
        counter = counter + 1;
        index(in) = counter;
    end
end


% Matrix Calculation
Nf = counter;
Sff = spalloc(Nf, Nf, 7*Nf);
Tff = spalloc(Nf, Nf, 7*Nf);
Tffc = spalloc(Nf, Nf, ceil(2*pi*sqrt(Nf)));
B = zeros(Nf, 1);


for ie = 1:Num_elements
    nodes(1:3) = t(1:3, ie);
    x(1:3) = p(1, nodes(1:3));
    y(1:3) = p(2, nodes(1:3));

    De = det([1 x(1) y(1); 1 x(2) y(2); 1 x(3) y(3)]);
    Ae = abs(De/2);
    b(1) = (y(2) - y(3))/De;
    b(2) = (y(3) - y(1))/De;
    b(3) = (y(1) - y(2))/De;
    c(1) = (x(3) - x(2))/De;
    c(2) = (x(1) - x(3))/De;
    c(3) = (x(2) - x(1))/De;

    for i = 1:3
        for j = 1:3
            Se(i, j) = (1/m0)*(b(i)*b(j) + c(i)*c(j))*Ae;
            if i == j
                Te(i, j) = e0*Ae/6;
            else
                Te(i, j) = e0*Ae/12;
            end

            if(node_id(nodes(i)) == 1)
                if(node_id(nodes(j)) ~= 0)
                    Sff(index(nodes(i)), index(nodes(j))) = Sff(index(nodes(i)), index(nodes(j))) + Se(i, j);
                    Tff(index(nodes(i)), index(nodes(j))) = Tff(index(nodes(i)), index(nodes(j))) - omega*Te(i, j);
                else
                    B(index(nodes(i))) = B(index(nodes(i))) - (Se(i, j) - omega*Te(i, j))*Ez(nodes(j));
                end
            end
            if(node_id(nodes(i)) == 2)
                if(node_id(nodes(j)) == 1)
                    Sff(index(nodes(i)), index(nodes(j))) = Sff(index(nodes(i)), index(nodes(j))) + Se(i, j);
                    Tff(index(nodes(i)), index(nodes(j))) = Tff(index(nodes(i)), index(nodes(j))) - omega*Te(i, j);
                elseif(node_id(nodes(j)) == 2)
                    Sff(index(nodes(i)), index(nodes(j))) = Sff(index(nodes(i)), index(nodes(j))) + Se(i, j);
                    Tff(index(nodes(i)), index(nodes(j))) = Tff(index(nodes(i)), index(nodes(j))) - omega*Te(i, j);
                    Tffc(index(nodes(i)), index(nodes(j))) = Tffc(index(nodes(i)), index(nodes(j))) + (1/(m0*e0))*alpha*Te(i,j);
                end
            end
        end
    end
end

A = Sff + Tff + Tffc;
el = A\B;
for in = 1:Num_nodes
    if (node_id(in) ~= 0)
        Ez(in) = abs(el(index(in))+exp(-1i*k*p(1,in))); 
     else
         Ez(in) = abs(Ez(in)+exp(-1i*k*p(1,in)));
    end
end

pdeplot(p, e, t, 'xydata', Ez);
axis tight;
axis equal;
hold on;
colormap jet;

File: Python/CircularCylinderScatteringFEM.py
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.tri as mtri
import scipy.sparse as sp

lam = 1
f = 300*pow(10,6)
k = (2*np.pi)/lam
a = lam/4
R1 = a + lam/2
E0 = 1
m0 = 1.256637061*pow(10,-6)
e0 = 8.854187817*pow(10,-12)  

alpha = (1/(2*R1) + 1j*k)
omega = pow(2*np.pi*f,2)


# Create Geometry
R = np.linspace(a, R1, 100)
Theta = np.linspace(0, 2*np.pi, 150)

R_matrix, Theta_matrix = np.meshgrid(R, Theta)
X_n = R_matrix * np.cos(Theta_matrix)
Y_n = R_matrix * np.sin(Theta_matrix)
X = X_n.reshape(X_n.size)
Y = Y_n.reshape(Y_n.size)

triang = mtri.Triangulation(X, Y)
tri_coords = np.zeros((triang.triangles.shape[0], 3, 2))
for i, tri in enumerate(triang.triangles):
    tri_coords[i] = np.column_stack((X[tri], Y[tri]))

mask = np.zeros(triang.triangles.shape[0], dtype=bool)
for i in range(0, triang.triangles.shape[0]):
    t_c = tri_coords[i]
    x1 = t_c[0][0]
    x2 = t_c[1][0]
    x3 = t_c[2][0]
    y1 = t_c[0][1]
    y2 = t_c[1][1]
    y3 = t_c[2][1]
    r1 = np.sqrt(pow(x1, 2) + pow(y1, 2))
    r2 = np.sqrt(pow(x2, 2) + pow(y2, 2))
    r3 = np.sqrt(pow(x3, 2) + pow(y3, 2))
    if r1-a <= pow(10,-6) and r2-a <= pow(10,-6) and r3-a <= pow(10,-6):
        mask[i] = True
triang.set_mask(mask)

nodes = np.column_stack((triang.x, triang.y))
unique_nodes = np.unique(nodes, axis=0)
element_nodes = triang.triangles


# Define known and unknown fields
node_id = np.ones(nodes.shape[0])
Ez = np.zeros(nodes.shape[0], dtype=complex)
for inode in range(0, nodes.shape[0]):
    radius = np.sqrt(pow(nodes[inode][0],2) + pow(nodes[inode][1], 2))
    if abs(radius-a) <= pow(10,-6):
        node_id[inode] = 0
        Ez[inode] = -E0*np.exp(-1j*k*nodes[inode][0])
    elif abs(radius-R1) <= pow(10,-6):
        node_id[inode] = 2

# Index of unkowns
un_index = np.zeros(nodes.size, dtype=int)
counter = 0
for inode in range(0, nodes.shape[0]):
    if node_id[inode] != 0:
        un_index[inode] = counter
        counter = counter + 1
      


# Matrix Calculation
Nf = counter
Sff = sp.lil_matrix((Nf, Nf), dtype=complex)
Tff = sp.lil_matrix((Nf, Nf), dtype=complex)
Tffc = sp.lil_matrix((Nf, Nf), dtype=complex)
B = np.zeros(Nf, dtype=complex)


for ie in range(1, triang.triangles.shape[0]):
    t_c = tri_coords[ie]
    x = np.zeros(3)
    y = np.zeros(3)
    x = [t_c[0][0], t_c[1][0], t_c[2][0]]
    y = [t_c[0][1], t_c[1][1], t_c[2][1]]
    n = [element_nodes[ie][0], element_nodes[ie][1], element_nodes[ie][2]]
        
    A = np.array([[1, x[0], y[0]], [1, x[1], y[1]], [1, x[2], y[2]]])
    De = np.linalg.det(A)
    Ae = abs(De/2)
    b = np.zeros(3)
    c = np.zeros(3)
    b = [(y[1]-y[2])/De, (y[2]-y[0])/De, (y[0]-y[1])/De]
    c = [(x[2]-x[1])/De, (x[0]-x[2])/De, (x[1]-x[0])/De]

    Se = np.zeros([3, 3], dtype=complex)
    Te = np.zeros([3, 3], dtype=complex)
    Tec = np.zeros([3, 3], dtype=complex)
    for i in range(0, 3):
        for j in range(0, 3):
            Se[i][j] = (1/m0)*(b[i]*b[j] + c[i]*c[j])*Ae
            if i == j:
                Te[i][j] = omega*e0*Ae/6
            else:
                Te[i][j] = omega*e0*Ae/12

            if node_id[n[i]] == 1:
                if node_id[n[j]] != 0:
                    Sff[un_index[n[i]], un_index[n[j]]] = Sff[un_index[n[i]], un_index[n[j]]] + Se[i][j]
                    Tff[un_index[n[i]], un_index[n[j]]] = Tff[un_index[n[i]], un_index[n[j]]] - Te[i][j]
                else:
                    B[un_index[n[i]]] = B[un_index[n[i]]] - (Se[i][j] - Te[i][j])*Ez[n[j]]

            if node_id[n[i]] == 2:
                if node_id[n[j]] == 1:
                    Sff[un_index[n[i]], un_index[n[j]]] = Sff[un_index[n[i]], un_index[n[j]]] + Se[i][j]
                    Tff[un_index[n[i]], un_index[n[j]]] = Tff[un_index[n[i]], un_index[n[j]]] - Te[i][j]
                elif node_id[n[j]] == 2:
                    Sff[un_index[n[i]], un_index[n[j]]] = Sff[un_index[n[i]], un_index[n[j]]] + Se[i][j]
                    Tff[un_index[n[i]], un_index[n[j]]] = Tff[un_index[n[i]], un_index[n[j]]] - Te[i][j]

                    M1 = [(x[i] + x[j])/2, (y[i] + y[j])/2]
                    M2 = [(x[j] + x[np.mod(j,3)])/2, (y[j] + y[np.mod(j,3)])/2]
                    AP = [x[np.mod(j,3)] - x[i], y[np.mod(j,3)] - y[i]]   
                    BP = [x[np.mod(i,3)] - x[j], y[np.mod(i,3)] - y[j]]
                    Si = np.linalg.norm(np.cross(np.append(M1, 0), np.append(AP, 0))) / 2
                    Sj = np.linalg.norm(np.cross(np.append(M2, 0), np.append(BP, 0))) / 2 
                    Tec[i, j] = (1/2)*(alpha/m0)*(2*Si + Sj)

                    Tffc[un_index[n[i]], un_index[n[j]]] = Tffc[un_index[n[i]], un_index[n[j]]] + Tec[i,j]
            


# Unkown Field Calculation
A = Sff + Tff + Tffc
eln = sp.linalg.bicgstab(A, B)
el = eln[0]
for inode in range(0, nodes.shape[0]):
    if node_id[inode] != 0:
        Ez[inode] = np.abs(el[un_index[inode]]+np.exp(-1j*k*nodes[inode][0]))
    else:
        Ez[inode] = np.abs(Ez[inode]+np.exp(-1j*k*nodes[inode][0]))

Ez = np.real(Ez)

# Plot the results
fig, ax = plt.subplots()
ax.triplot(triang)
cax = ax.tripcolor(triang, Ez, cmap='plasma', shading='flat')
fig.colorbar(cax, ax=ax)
ax.set_xlabel('x')
ax.set_ylabel('y')
plt.show()